### PR TITLE
[search] LocalityScorer: Prefer nearest locality.

### PR DESCRIPTION
**Pull request by tatiana-yan**

### CODE REVIEW

**Comment by tatiana-yan** [comment]
JTMAC

**Comment by tatiana-yan** [comment]
JTMAC

**Comment by tatiana-yan** [comment]
JTMAC

File: search/locality_scorer.hpp [hunk 1]
```diff
@@ -30,14 +32,17 @@ class LocalityScorer
 
     virtual void GetNames(uint32_t featureId, std::vector<std::string> & names) const = 0;
     virtual uint8_t GetRank(uint32_t featureId) const = 0;
+    virtual std::optional<m2::PointD> GetCenter(uint32_t featureId) = 0;
   };
 
-  LocalityScorer(QueryParams const & params, Delegate const & delegate);
+  LocalityScorer(QueryParams const & params, m2::PointD const & pivot, Delegate & delegate);
+
+  void SetPivotForTests(m2::PointD const & pivot) { m_pivot = pivot; }
```
Comment (mpimenov) [inline search/locality_scorer.hpp:40]: nit: ForTesting is more common than ForTests in search/

File: search/locality_scorer.hpp [hunk 2]
```diff
@@ -50,27 +55,31 @@ class LocalityScorer
...
     uint8_t m_rank = 0;
+    double m_distanceToPivot = std::numeric_limits<double>::max();
   };
 
   friend std::string DebugPrint(ExLocality const & locality);
 
   // Leaves at most |limit| elements of |localities|, ordered by some
   // combination of ranks and number of matched tokens.
-  void LeaveTopLocalities(IdfMap & idfs, size_t limit, std::vector<Locality> & localities) const;
+  void LeaveTopLocalities(IdfMap & idfs, size_t limit, std::vector<Locality> & localities);
```
Comment (mpimenov) [inline search/locality_scorer.hpp:65]: documentation

File: search/search_tests/locality_scorer_test.cpp [hunk 3]
```diff
@@ -263,3 +282,27 @@ UNIT_CLASS_TEST(LocalityScorerTest, Similarity)
...
+    ID_ABERDEEN_CLOSE,
+    ID_ABERDEEN_RANK1,
+    ID_ABERDEEN_RANK2,
+    ID_ABERDEEN_RANK3
+  };
+
+  AddLocality("Aberdeen", ID_ABERDEEN_CLOSE, 10 /* rank */, m2::PointD(11.0, 11.0));
+  AddLocality("Aberdeen", ID_ABERDEEN_RANK1, 100 /* rank */, m2::PointD(0.0, 0.0));
+  AddLocality("Aberdeen", ID_ABERDEEN_RANK2, 50 /* rank */, m2::PointD(2.0, 2.0));
+  AddLocality("Aberdeen", ID_ABERDEEN_RANK2, 5 /* rank */, m2::PointD(4.0, 4.0));
```
Comment (mpimenov) [inline search/search_tests/locality_scorer_test.cpp:299]: s/RANK2/RANK3/

File: search/search_tests/locality_scorer_test.cpp [hunk 4]
```diff
@@ -263,3 +282,27 @@ UNIT_CLASS_TEST(LocalityScorerTest, Similarity)
...
+
+  InitParams("Aberdeen", m2::PointD(10.0, 10.0) /* pivot */, false /* lastTokenIsPrefix */);
+
+  // Expected order is: the closest one (ID_ABERDEEN_CLOSE) first, than sorted by rank.
```
Comment (mpimenov) [inline search/search_tests/locality_scorer_test.cpp:303]: s/than/then/

File: search/locality_scorer.hpp (hunk 1 above)
Comment (tatiana-yan) [reply search/locality_scorer.hpp]: поменяла

File: search/locality_scorer.hpp (hunk 2 above)
Comment (tatiana-yan) [reply search/locality_scorer.hpp:67]: переписала

File: search/search_tests/locality_scorer_test.cpp (hunk 3 above)
Comment (tatiana-yan) [reply search/search_tests/locality_scorer_test.cpp]: поправила

File: search/search_tests/locality_scorer_test.cpp (hunk 4 above)
Comment (tatiana-yan) [reply search/search_tests/locality_scorer_test.cpp]: поправила

**Comment by tatiana-yan** [comment]
JTMAC

File: search/locality_scorer.hpp [hunk 5]
```diff
@@ -60,8 +60,10 @@ class LocalityScorer
 
   friend std::string DebugPrint(ExLocality const & locality);
 
-  // Leaves at most |limit| elements of |localities|, ordered by some
-  // combination of ranks and number of matched tokens.
+  // Leaves at most |limit| elements of |localities|. Candidates are selected with
+  // LeaveTopByExactMatchNormAndRank. Number of candidates typically is match bigger than |limit|
```
Comment (mpimenov) [inline search/locality_scorer.hpp:64]: much?

Comment (tatiana-yan) [reply search/locality_scorer.hpp]: да, спасибо

**Comment by mpimenov** [comment]
JTMAC

**Pull request merged**